firedrake/slate/slac.py

```diff
@@ -0,0 +1,512 @@
+        :template: result `Eigen::MatrixBase<Derived>`.
+
+        i.e. double A[n][m] ---> const Eigen::MatrixBase<Derived> &A_
+        """
+
+        name = kwargs.get("name", "A")
+        if o.sym.symbol != name:
+            return o
+        newtype = "const Eigen::MatrixBase<Derived> &"
+
+        return o.reconstruct(newtype, ast.Symbol("%s_" % name))
+
+    def visit_Symbol(self, o, *args, **kwargs):
+        """Visits a COFFEE symbol and redefines it as a FunCall object.
+
+        i.e. A[j][k] ---> A(j, k)
+        """
+
+        name = kwargs.get("name", "A")
+        if o.symbol != name:
+            return o
+        shape = o.rank
+
+        return ast.FunCall(ast.Symbol(name), *shape)
+
+
+def compile_slate_expression(slate_expr, tsfc_parameters=None):
+    """Takes a SLATE expression `slate_expr` and returns the
+    appropriate :class:`firedrake.op2.Kernel` object representing
+    the SLATE expression.
+
+    :arg slate_expr: a :class:'TensorBase' expression.
+
+    :arg tsfc_parameters: an optional `dict` of form compiler parameters to
+                          be passed onto TSFC during the compilation of
+                          integral forms.
+    """
+    # Only SLATE expressions are allowed as inputs
+    if not isinstance(slate_expr, TensorBase):
+        raise ValueError("Expecting a `slate.TensorBase` expression, not a %r" % slate_expr)
+
+    # SLATE currently does not support arguments in mixed function spaces
+    # TODO: Get PyOP2 to write into mixed dats
+    if any(len(a.function_space()) > 1 for a in slate_expr.arguments()):
+        raise NotImplementedError("Compiling mixed slate expressions")
+
+    # Initialize variables: dtype and dictionary objects.
+    dtype = SCALAR_TYPE
+    shape = slate_expr.shape
+    temps = {}
+    kernel_exprs = {}
+    statements = []
+
+    # Extract all coefficents in the SLATE expression and construct the
+    # appropriate coefficent mapping
+    coeffs = slate_expr.coefficients()
+    coeffmap = dict((c, ast.Symbol("w%d" % i)) for i, c in enumerate(coeffs))
+
+    # By default, we don't assume that we need to perform facet integral loops.
+    # If an expression contains facet integrals within the expression, we activate
+    # this to indicate that we need to loop over cell facets later on.
+    need_cell_facets = False
+
+    # initialize coordinate and facet symbols
+    coordsym = ast.Symbol("coords")
+    coords = None
+    cellfacetsym = ast.Symbol("cell_facets")
+    inc = []
+
+    # Provide the SLATE expression as input to extract kernel functions
+    get_kernel_expr(slate_expr, tsfc_parameters, temps, kernel_exprs, statements)
+
+    # Now we construct the body of the macro kernel
+    for exp, t in temps.items():
+        statements.append(ast.FlatBlock("%s.setZero();\n" % t))
+
+        for klist in kernel_exprs[exp]:
+            clist = []
+            index = klist[0]
+            kinfo = klist[1]
```

Comment:
When you have named tuples, please use names instead of indices. It is easier to read (what is 0? what is 1?), and more robust to change. (Imagine someone in the future needs to add another member to this tuple, and for some reason doesn't want to put that new member last. Good luck finding all the random indices all over that need updating.)